### PR TITLE
Introduce a tutorial about fetching data using `package:http`

### PR DESCRIPTION
**Pull request by parlough**
This replaces the old Dart [Fetch data dynamically](https://dart.dev/tutorials/web/fetch-data) article which is web-based and not platform independent.

This PR introduces a completely new tutorial built around `package:http` allowing the logic to work to work on multiple platforms, including for CLI, web, and Flutter apps.

The tutorial uses mock package information data for its examples, hosted directly on the site.

I haven't written a full tutorial on the sites before so please be harsh!

**Staged:** https://dart-dev--pr4520-feature-package-http-lfma2l8x.web.app/tutorials/server/fetch-data

Closes https://github.com/dart-lang/site-www/issues/3852

### CODE REVIEW

**Comment by parlough** [comment]
Not a priority, but could you review this when you get a chance @domesticmouse? I appreciate your feedback :)

**Review by domesticmouse — APPROVED** [review]
My high level question is should we be guiding developers to `dart:convert` or should we be guiding them to `freezed` and `built_value`? I'd argue that's probably a more sane approach.

With that out of the way, the content looks good. =)

**Comment by parlough** [comment]
@domesticmouse Yeah I'd love to direct users in that direction too, but thought it was a bit of a more orthogonal concept from fetching data and wanted to keep these steps simpler and able to build in DartPad (so without code generation). 

My follow-up plan is to fix https://github.com/dart-lang/site-www/issues/4286 and expand on our dart.dev JSON documentation and then link to there more prominently on this page in the JSON section and as a next step. How does that sound?

**Comment by domesticmouse** [comment]
> My follow-up plan is to fix #4286 and expand on our dart.dev JSON documentation and then link to there more prominently on this page in the JSON section and as a next step. How does that sound?

Sounds good to me

File: src/_tutorials/server/fetch-data.md [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+
+You can directly use `dart:io` or `dart:html` to make HTTP requests,
+however those libraries are platform dependent.
+`package:http` provides a cross-platform library
+for making composable HTTP requests,
+with optional fine-grained control.
+
+To add a dependency on `package:http`,
+run the [`dart pub add`][] command
+while specifying `http`:
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:133]:
```suggestion
run the following [`dart pub add`][] command
from the top of your repo:
```

File: src/_tutorials/server/fetch-data.md [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+
+```terminal
+$ dart pub add http
+```
+
+To then use `package:http` in your code,
+import it and optionally [specify a library prefix][]:
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:140]:
```suggestion
To use `package:http` in your code,
import it and optionally [specify a library prefix][]:
```

File: src/_tutorials/server/fetch-data.md [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+and its [API documentation][http-docs].
+
+[`dart pub add`]: /tools/pub/cmd/pub-add
+[specify a library prefix]: /guides/language/language-tour#specifying-a-library-prefix
+
+## Build a URL
+
+As previously mentioned,
+to make an HTTP request,
+you first need a URL which identifies
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:158]:
```suggestion
you first need a URL that identifies
```

File: src/_tutorials/server/fetch-data.md [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+[`Uri`]: {{site.dart-api}}/dart-core/Uri-class.html
+[library-tour-uri]: /guides/libraries/library-tour#uris
+
+## Make a network request
+
+If you just need to quickly fetch a string representation
+of a requested resource,
+you can use the top-level [`read`][http-read]
+function found in `package:http`
+which returns a `Future<String>` or throws
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:194]:
```suggestion
that returns a `Future<String>` or throws
```

File: src/_tutorials/server/fetch-data.md [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+  "latestVersion": "0.13.5",
+  "description": "A composable, multi-platform, Future-based API for HTTP requests.",
+  "publisher": "dart.dev",
+  "repository": "https://github.com/dart-lang/http"
+}
+```  
+
+Note the structure of the data,
+in this case a map,
+as you will need it when decoding the JSON later on.
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:235]:
```suggestion
Note the structure of the data
(in this case a `Map`),
as you will need it when decoding the JSON later on.
```

File: src/_tutorials/server/fetch-data.md [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+
+If you need other information from the response,
+such as the [status code][] or the [headers][],
+you can instead use the top-level [`get`][http-get] function
+which returns a `Future` with a [`Response`][http-response].
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:240]:
```suggestion
that returns a `Future` with a [`Response`][http-response].
```

File: src/_tutorials/server/fetch-data.md [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+  if (httpPackageResponse.statusCode != 200) {
+    print('Failed to retrieve the http package!');
+    return;
+  }
+  print(httpPackageResponse.body);
+}
+```
+
+There are many other status codes besides **200**
+and your app may want to handle them differently.
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:260]:
```suggestion
and your app might want to handle them differently.
```

File: src/_tutorials/server/fetch-data.md [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+To learn more about what different status codes mean,
+see [HTTP response status codes][] on the mdn web docs.
+
+If the endpoint you are requesting from requires more information,
+such as authentication or user-agent information,
+it often requires you to include [HTTP headers][headers].
+You can specify headers by passing in a `Map<String, String>`
+of the key-value pairs to the `headers` optional named parameter:
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:268]:
```suggestion
Some server requests require more information,
such as authentication or user-agent information;
in this case you might need to include [HTTP headers][headers].
You can specify headers by passing in a `Map<String, String>`
of the key-value pairs as the `headers` optional named parameter:
```

File: src/_tutorials/server/fetch-data.md [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+    client.close();
+  }
+}
+```
+
+The `RetryClient` has a default behavior
+for how many times to retry and how long between each request,
+but its behavior can be modified through parameters
+to the [`RetryClient()`][http-retry-client-cons]
+or [`RetryClient.withDelays()`][http-retry-client-delay] constructors.
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:334]:
```suggestion
The `RetryClient` has a default behavior for how many
times to retry and how long to wait between each request,
but its behavior can be modified through parameters
to the [`RetryClient()`][http-retry-client-cons]
or [`RetryClient.withDelays()`][http-retry-client-delay] constructors.
```

File: src/_tutorials/server/fetch-data.md [hunk 10]
```diff
@@ -0,0 +1,631 @@
...
+accessing specific portions of information
+from a string can be a challenge.
+
+Since the data is already in a JSON format,
+you can use Dart's built-in [`json.decode`][decode-docs] function
+in the `dart:convert` library
+to convert the raw string into
+a JSON representation using Dart objects.
+In this case, the JSON data is represented in a map structure
+and in JSON, map keys are always strings,
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:359]:
```suggestion
and, in JSON, map keys are always strings,
```

File: src/_tutorials/server/fetch-data.md [hunk 11]
```diff
@@ -0,0 +1,631 @@
...
+}
+```
+
+[decode-docs]: {{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-convert/JsonCodec/decode.html
+
+### Create a structured class to store the data
+
+To provide the decoded JSON with more structure,
+making it easier to work with,
+you can create a class which can store the
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:382]:
```suggestion
create a class that can store the
```

File: src/_tutorials/server/fetch-data.md [hunk 12]
```diff
@@ -0,0 +1,631 @@
...
+retrieved data using specific types depending
+on the schema of your data.
+
+The following snippets shows a class-based representation
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:386]:
```suggestion
The following snippet shows a class-based representation
```

File: src/_tutorials/server/fetch-data.md [hunk 13]
```diff
@@ -0,0 +1,631 @@
...
+which can store the package information returned
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:387]:
```suggestion
that can store the package information returned
```

File: src/_tutorials/server/fetch-data.md [hunk 14]
```diff
@@ -0,0 +1,631 @@
...
+      latestVersion: json['latestVersion'] as String,
+      description: json['description'] as String,
+      publisher: json['publisher'] as String,
+      repository: repository != null ? Uri.tryParse(repository) : null,
+    );
+  }
+}
+```
+
+A handwritten method, such as used here,
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:442]:
```suggestion
A handwritten method, such the previous example,
```

File: src/_tutorials/server/fetch-data.md [hunk 15]
```diff
@@ -0,0 +1,631 @@
...
+To learn more about JSON serialization and deserialization,
+including automatic generation of the conversion logic,
+see the [Using JSON][] guide.
+
+### Convert the response to an object of your structured class
+
+Now you have a class to store your data
+and a way to convert the decoded JSON object
+into an object of that type.
+Next you can write a function which
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:454]:
```suggestion
Next, you can write a function that
```
https://writeanything.wordpress.com/2008/09/20/grammar-girl-ie-vs-eg/

File: src/_tutorials/server/fetch-data.md [hunk 16]
```diff
@@ -0,0 +1,631 @@
...
+puts everything to together:
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:455]:
```suggestion
pulls everything together:
```

File: src/_tutorials/server/fetch-data.md [hunk 17]
```diff
@@ -0,0 +1,631 @@
...
+
+Now that you've retrieved data and
+converted it to a more easily accessible format,
+you can use it however you'd like.
+Some possibilities include
+outputting information to a CLI, or
+displaying it in a [web][] or [Flutter][] app.
+
+Here is complete, runnable example
+which requests, decodes, then displays
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:503]:
```suggestion
that requests, decodes, then displays
```

File: src/_tutorials/server/fetch-data.md [hunk 18]
```diff
@@ -0,0 +1,631 @@
...
+
+    buf.write('!');
+    return buf.toString();
+  }
+}
+```
+
+{{site.alert.flutter-note}}
+  For another example that covers fetching then displaying data in Flutter,
+  see the [Fetching data from the internet][] Flutter cookbook.
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:607]:
```suggestion
  see the [Fetching data from the internet][] Flutter recipe.
```

File: src/_tutorials/server/fetch-data.md [hunk 19]
```diff
@@ -0,0 +1,631 @@
...
+{{site.alert.end}}
+
+[web]: /web
+[Flutter]: {{site.flutter}}
+[Fetching data from the internet]: {{site.flutter-docs}}/cookbook/networking/fetch-data
+
+## What next?
+
+Now that you have retrieved, parsed, and utilized
```
Comment (sfshaza2) [inline src/_tutorials/server/fetch-data.md:616]:
```suggestion
Now that you have retrieved, parsed, and used
```

**Review by sfshaza2 — APPROVED** [review]
Great job, @parlough! I found very minor nits. Mostly, you might want to read that link I provided from the Grammar girl where she explains "that" vs "which".  Once you deal with the minor corrections, lgtm!

**Comment by parlough** [comment]
Thanks @sfshaza2! I appreciate the fixes, feedback, and context :)

**Review by MaryaBelanger — APPROVED** [review]
This is _great_. I could follow all the way through (despite having no background knowledge on the topic; nice background section!), and I really like the progression, it just _makes sense_. 

Even just scanning the headings, I know how this works and the things I'd need to do to accomplish it, without having to do anything. In that way, it works for beginners, who can follow the details of each section to apply it to their use case, and advanced readers, who can scan the headings and the code snippets and get exactly what they need. 

Like we've been talking about, there'll be some type-of-info restructuring across the site in the future that might affect this slightly (like maybe the background info per topic will move to the Glossary we discussed, or have their own conceptual pages, and then be linked to from the tutorial). But all in all, this is a great, clean structure that I think we'll be using to define a standard for organization and layout of future tutorials. And hopefully having that standard well-defined will make it easier for many more external users to contribute their own tutorials. 

Thank you very much @parlough! Creating content while simultaneously improving _how_ that content is created is not easy, and extremely valuable!

**Comment by parlough** [comment]
Thanks so much for your thoughts @MaryaBelanger! :D

I really appreciate your perspective as you went through the tutorial as it seems I was able to achieve the goals I was aiming for, particularly with the headings and how I organized the different sections. 

I'm looking forward to working towards a well-defined structure together as we work on those new features and restructuring and I'm glad this seems to be heading in an appropriate direction!

**Comment by parlough** [comment]
I'm going to merge this with the review from all three of you (thanks so much!), but will be on the look out for more feedback once released. 🥳

**Pull request merged**